lxd/identities.go

```diff
@@ -166,6 +174,358 @@ func identityAccessHandler(authenticationMethod string, entitlement auth.Entitle
+			return response.InternalError(err)
+		}
+
+		// Generate fingerprint of network certificate so joining member can automatically trust the correct
+		// certificate when it is presented during the join process.
+		fingerprint, err := shared.CertFingerprintStr(string(s.Endpoints.NetworkPublicKey()))
+		if err != nil {
+			return response.InternalError(err)
+		}
+
+		if req.Groups == nil {
+			req.Groups = []string{}
+		}
+
+		expiry := s.GlobalConfig.RemoteTokenExpiry()
+		var expiresAt time.Time
+		if expiry != "" {
+			expiresAt, err = shared.GetExpiry(time.Now(), expiry)
+			if err != nil {
+				return response.InternalError(err)
+			}
+		}
+
+		identifier := uuid.New()
+		metadata := dbCluster.PendingTLSMetadata{
+			Secret: joinSecret,
+			Expiry: expiresAt,
+		}
+
+		metadataJSON, err := json.Marshal(metadata)
+		if err != nil {
+			return response.InternalError(fmt.Errorf("Failed to encode pending TLS identity metadata: %w", err))
+		}
+
+		err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+			id, err := dbCluster.CreateIdentity(ctx, tx.Tx(), dbCluster.Identity{
+				AuthMethod: api.AuthenticationMethodTLS,
+				Type:       api.IdentityTypeCertificateClientPending,
+				Identifier: identifier.String(),
+				Name:       req.Name,
+				Metadata:   string(metadataJSON),
+			})
+			if err != nil {
+				return err
+			}
+
+			return dbCluster.SetIdentityAuthGroups(ctx, tx.Tx(), int(id), req.Groups)
+		})
+		if err != nil {
+			return response.InternalError(fmt.Errorf("Failed to create pending TLS identity: %w", err))
+		}
+
+		token := api.CertificateAddToken{
+			ClientName:  req.Name,
+			Fingerprint: fingerprint,
+			Addresses:   addresses,
+			Secret:      joinSecret,
+			ExpiresAt:   expiresAt,
+			Type:        api.IdentityTypeCertificateClient,
+		}
+
+		lc, err := notify(lifecycle.IdentityCreated, identifier.String(), false)
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		return response.SyncResponseLocation(true, token, lc.Source)
+	}
+
+	if !trusted {
+		// Check if certificate add token supplied as token.
+		joinToken, err := shared.CertificateTokenDecode(req.TrustToken)
+		if err != nil {
+			return response.Forbidden(nil)
+		}
+
+		// If so then check there is a matching pending TLS identity.
+		identifier, err := tlsIdentityTokenValid(r.Context(), s, *joinToken)
+		if err != nil {
+			return response.InternalError(fmt.Errorf("Failed during search for pending TLS identity: %w", err))
+		}
+
+		err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+			return dbCluster.ActivateTLSIdentity(ctx, tx.Tx(), identifier, cert)
+		})
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		lc, err := notify(lifecycle.IdentityUpdated, identifier.String(), true)
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		return response.SyncResponseLocation(true, nil, lc.Source)
+	}
+
+	// Calculate the fingerprint.
```

Comment:
This section could be a separate function dealing with creating identity directly